src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.search.collector;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Locale;
+import java.util.ArrayList;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.FieldValueHitQueue;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.search.TopFieldDocs;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.FieldDoc;
+import org.apache.lucene.util.PriorityQueue;
+import org.opensearch.neuralsearch.query.HybridQueryScorer;
+import org.opensearch.common.Nullable;
+import org.opensearch.neuralsearch.search.HitsThresholdChecker;
+import org.opensearch.neuralsearch.search.lucene.MultiLeafFieldComparator;
+
+/*
+ Collects the TopFieldDocs after executing hybrid query. Uses HybridQueryTopDocs as DTO to handle each sub query results.
+ The individual query results are sorted as per the sort criteria sent in the search request.
+ */
+@Log4j2
+public abstract class HybridTopFieldDocSortCollector implements Collector {
+    private final int numHits;
+    private final HitsThresholdChecker hitsThresholdChecker;
+    private final Sort sort;
+    private FieldComparator<?> firstComparator;
+    private FieldValueHitQueue.Entry bottom = null;
+    @Getter
+    private int totalHits;
+    protected int docBase;
+    protected LeafFieldComparator comparators[];
+    @Getter
+    @Setter
+    private TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
+    /*
+      reverseMul is used to set the direction of the sorting when creating comparators.
+      In threshold check reverseMul is used in comparison logic.
+      It modifies the comparison of either reverse or maintain the natural order depending on its value.
+      This ensures that the compareBottom method adjusts the order based on whether you want ascending or descending sorting.
+     */
+    protected int reverseMul;
+    protected FieldValueHitQueue<FieldValueHitQueue.Entry>[] compoundScores;
+    protected boolean queueFull[];
+    @Getter
+    protected float maxScore = 0.0f;
+    protected int[] collectedHits;
+
+    // searchSortPartOfIndexSort is used to evaluate whether to perform index sort or not.
+    private Boolean searchSortPartOfIndexSort = null;
```

Comment:
The reason why it is kept as null because we need to make sure that for every shard this value has to be **null** and to avoid having any garbage value this condition is a good to keep.